packages/cw721/src/state.rs

```diff
@@ -0,0 +1,460 @@
+    for CollectionMetadataMsg<TCollectionMetadataExtensionMsg>
+where
+    TCollectionMetadataExtension: Cw721State,
+    TCollectionMetadataExtensionMsg: Cw721CustomMsg + StateFactory<TCollectionMetadataExtension>,
+{
+    fn create(
+        &self,
+        deps: Deps,
+        env: &cosmwasm_std::Env,
+        info: &cosmwasm_std::MessageInfo,
+        current: Option<&CollectionMetadata<TCollectionMetadataExtension>>,
+    ) -> Result<CollectionMetadata<TCollectionMetadataExtension>, Cw721ContractError> {
+        self.validate(deps, env, info, current)?;
+        match current {
+            // Some: update existing metadata
+            Some(current) => {
+                let mut updated = current.clone();
+                if let Some(name) = &self.name {
+                    updated.name = name.clone();
+                }
+                if let Some(symbol) = &self.symbol {
+                    updated.symbol = symbol.clone();
+                }
+                let current_extension = current.extension.clone();
+                let updated_extension =
+                    self.extension
+                        .create(deps, env, info, Some(&current_extension))?;
+                updated.extension = updated_extension;
+                Ok(updated)
+            }
+            // None: create new metadata
+            None => {
+                let extension = self.extension.create(deps, env, info, None)?;
+                let new = CollectionMetadata {
+                    name: self.name.clone().unwrap(),
+                    symbol: self.symbol.clone().unwrap(),
+                    extension,
+                    updated_at: env.block.time,
+                };
+                Ok(new)
+            }
+        }
+    }
+
+    fn validate(
+        &self,
+        deps: Deps,
+        _env: &cosmwasm_std::Env,
+        info: &cosmwasm_std::MessageInfo,
+        _current: Option<&CollectionMetadata<TCollectionMetadataExtension>>,
+    ) -> Result<(), Cw721ContractError> {
+        // collection metadata can only be updated by the creator
+        // - case 1: skip in case of init, since there is no creator yet
+        let creator_initialized = CREATOR.item.may_load(deps.storage)?;
+        // - case 2: check if sender is creator
+        if (self.name.is_some() || self.symbol.is_some())
+            && creator_initialized.is_some()
+            && CREATOR.assert_owner(deps.storage, &info.sender).is_err()
+        {
+            return Err(Cw721ContractError::NotCollectionCreator {});
+        }
+        // make sure the name and symbol are not empty
+        if self.name.is_some() && self.name.clone().unwrap().is_empty() {
+            return Err(Cw721ContractError::CollectionNameEmpty {});
+        }
+        if self.symbol.is_some() && self.symbol.clone().unwrap().is_empty() {
+            return Err(Cw721ContractError::CollectionSymbolEmpty {});
+        }
+        Ok(())
```

Comment:
... here in `validate()` in makes sure that `name` and `symbol` in `CollectionMetadata` is not empty.